### PR TITLE
Feature/508 create datastore via dialog

### LINKED ISSUE #508: Access datastore management via a dialog
Often times when configuring components that take a datastore as part of their configuration (such as table lookup, insert into table etc.) it is inconvenient that you cannot register a _new_ datastore at that time. Then you have to open a whole new window and go to datastore mgmt. etc.

Rather we should have a button next to those datastore comboboxes and when clicking it the user will have the datastore mgmt. panel showing in a popup or dialog that he can discard when finished with it.

Tweaks compared to the datastore mgmt _panel_:

 * There should not be a button for building a new job.
 * No checkboxes to select any specific datastore.
 * No "welcome panel" header. Instead a normal banner as used in other dialogs.

### PR DESCRIPTION
**Pull request by tomaszguzialek**
Fixes #508 

Not happy about duplicating the code from DatastoreManagementPanel, but there the list of available datastores is built manually.

### CODE REVIEW

File: desktop/api/src/main/java/org/datacleaner/util/WidgetFactory.java [hunk 1]
```diff
@@ -116,6 +116,21 @@ public static PopupButton createDefaultPopupButton(String text, String imagePath
         return b;
     }
 
+    public static PopupButton createSmallPopupButton(String text, String imagePath) {
+        PopupButton b = new PopupButton(text, ImageManager.get().getImageIcon(imagePath, IconUtils.ICON_SIZE_SMALL));
```
Comment (LosD) [inline desktop/api/src/main/java/org/datacleaner/util/WidgetFactory.java]: Missing final

File: desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java [hunk 2]
```diff
@@ -88,7 +147,249 @@ public void onItemSelected(Datastore item) {
...
+        populateCreateDatastoreMenu(menu);
+    }
+
+    private void populateCreateDatastoreMenu(JPopupMenu createDatastoreMenu) {
+        List<String> alreadyAddedDatabases = new ArrayList<>();
+
+        final JMenuItem csvMenuItem = WidgetFactory.createMenuItem("CSV file", IconUtils.CSV_IMAGEPATH);
+        csvMenuItem.addActionListener(createActionListener(CsvDatastore.class, CsvDatastoreDialog.class));
+        createDatastoreMenu.add(csvMenuItem);
+        alreadyAddedDatabases.add("CSV file");
```
Comment (LosD) [inline desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java]: Yeah, these are pretty nasty... Let's keep it as-is for now, but I think we have a little technical debt here.

File: desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java [hunk 3]
```diff
@@ -88,7 +147,249 @@ public void onItemSelected(Datastore item) {
         Datastore currentValue = (Datastore) componentBuilder.getConfiguredProperty(propertyDescriptor);
         setValue(currentValue);
 
-        add(_comboBox);
+        _createDatastoreButton = WidgetFactory.createSmallPopupButton("",
+                IconUtils.ACTION_CREATE_TABLE);
+        _createDatastoreButton.setToolTipText("Create datastore");
+        JPopupMenu createDatastoreMenu = _createDatastoreButton.getMenu();
```
Comment (LosD) [inline desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java]: final

File: desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java [hunk 4]
```diff
@@ -88,7 +147,249 @@ public void onItemSelected(Datastore item) {
...
+    
+    public void setOnlyUpdatableDatastores(boolean onlyUpdatableDatastores) {
+        _onlyUpdatableDatastores = onlyUpdatableDatastores;
+        final JPopupMenu menu = _createDatastoreButton.getMenu();
+        menu.removeAll();
+        populateCreateDatastoreMenu(menu);
+    }
+
+    private void populateCreateDatastoreMenu(JPopupMenu createDatastoreMenu) {
+        List<String> alreadyAddedDatabases = new ArrayList<>();
```
Comment (LosD) [inline desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java]: final

File: desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java [hunk 5]
```diff
@@ -88,7 +147,249 @@ public void onItemSelected(Datastore item) {
...
+        }
+
+        if (!_onlyUpdatableDatastores) {
+            final JMenuItem hbaseMenuItem = WidgetFactory.createMenuItem("HBase database", IconUtils.HBASE_IMAGEPATH);
+            hbaseMenuItem.addActionListener(createActionListener(HBaseDatastore.class, HBaseDatastoreDialog.class));
+            createDatastoreMenu.add(hbaseMenuItem);
+            alreadyAddedDatabases.add("HBase database");
+        }
+
+        DatabaseDriverCatalog databaseDriverCatalog = _dcModule.createInjectorBuilder().getInstance(
```
Comment (LosD) [inline desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java]: final

File: desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java [hunk 6]
```diff
@@ -88,7 +147,249 @@ public void onItemSelected(Datastore item) {
...
+            }
+
+        };
+    }
+
+    private ActionListener createJdbcActionListener(final String databaseName) {
+        return new ActionListener() {
+            @Override
+            public void actionPerformed(ActionEvent event) {
+                Injector injectorWithDatastore = _dcModule.createInjectorBuilder().with(JdbcDatastore.class, null)
```
Comment (LosD) [inline desktop/ui/src/main/java/org/datacleaner/widgets/properties/SingleDatastorePropertyWidget.java]: final

**Comment by LosD** [comment]
Except for the duplication, there's a few missing finals, but nothing else.

**Comment by kaspersorensen** [comment]
Hmm now we have three places in the code which brings up datastore dialogs (datastore mgmt, the buttons under the file dropzone and now this new dialog).

Could we at least create a class which expose some basic data objects containing datastore names, associated dialogs, icon etc.?

**Comment by tomaszguzialek** [comment]
I agree. I will create an issue for it.

**Pull request merged**